## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

In [1]:
import pandas as pd# importing pandas


In [2]:
df = pd.read_csv('/datasets/movies_and_shows.csv')# reading the files and storing them to df


In [3]:
df_shows = df.head(10)# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()


In [4]:
import pandas as pd
df = pd.read_csv('/datasets/movies_and_shows.csv')
print(df.info())# obtaining general information about the data in df


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB
None


The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [5]:
print(df.columns)# the list of column names in the df table


Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')


In [6]:
df = df.rename(
    columns = {'   name': 'name', 'Character': 'character', 'r0le': 'role', 'TITLE': 'title', '  Type': 'type', 'release Year': 'release_year', 'imdb sc0re': 'imdb_score', 'imdb v0tes': 'imdb_votes'})# renaming columns


In [7]:
print(df.columns)# checking result: the list of column names


Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')


### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [8]:
print(df.isna().sum())# calculating missing values


name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64


Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [9]:
df = df.dropna(axis='rows')
print(df)# dropping rows where columns with title, scores and votes have missing values


                     name                character      role        title  \
0          Robert De Niro            Travis Bickle     ACTOR  Taxi Driver   
1            Jodie Foster            Iris Steensma     ACTOR  Taxi Driver   
2           Albert Brooks                      Tom     ACTOR  Taxi Driver   
3           Harvey Keitel  Matthew 'Sport' Higgins     ACTOR  Taxi Driver   
4         Cybill Shepherd                    Betsy     ACTOR  Taxi Driver   
...                   ...                      ...       ...          ...   
85574    Adelaida Buscato               Mar??a Paz     ACTOR      Lokillo   
85575  Luz Stella Luengas             Karen Bayona     ACTOR      Lokillo   
85576        In??s Prieto                    Fanny     ACTOR      Lokillo   
85577        Isabel Gaona                   Cacica     ACTOR      Lokillo   
85578      Julian Gaviria                  unknown  DIRECTOR      Lokillo   

            type  release_year              genres  imdb_score  imdb_votes 

In [10]:
print(df.isna().sum())# counting missing values


name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64


### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [11]:
duplicates = df.duplicated()
print(duplicates.sum())# counting duplicate rows


6994


In [12]:
print(duplicates.tail())# Produce table with duplicates (with original rows included) and review last 5 rows


85574    False
85575    False
85576    False
85577     True
85578    False
dtype: bool


In [13]:
df = df.drop_duplicates()# removing duplicate rows


In [14]:
duplicates = df.duplicated()
print(duplicates)# checking for duplicates


0        False
1        False
2        False
3        False
4        False
         ...  
85573    False
85574    False
85575    False
85576    False
85578    False
Length: 73859, dtype: bool


In [15]:
sorted_df = df.sort_values(by='type')
unique_values = sorted_df['type'].unique()
print(unique_values)# viewing unique type names


['MOVIE' 'SHOW' 'movies' 'shows' 'the movie' 'tv' 'tv series' 'tv show'
 'tv shows']


In [16]:
def replace_wrong_show(wrong_shows_list, correct_show):
    for wrong_name in wrong_shows_list:
        df['type'] = df['type'].replace(wrong_name, correct_show)# function for replacing implicit duplicates


In [17]:
wrong_shows_list = ['shows', 'SHOW', 'tv show', 'tv shows', 'tv series', 'tv']
correct_show = 'show'
replace_wrong_show(wrong_shows_list, correct_show)# removing implicit duplicates


In [18]:
df['type'].unique()# viewing unique genre names


array(['MOVIE', 'the movie', 'show', 'movies'], dtype=object)

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

In [19]:
new_shows_df = df[df['release_year'] >= 1999]# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
new_name_df = new_shows_df.copy()# give the slice of dataframe new name


In [20]:
df_shows = new_shows_df[new_shows_df['type'] == 'show']# repeat conditional indexing so df has only shows (movies are removed as result)


In [21]:
new_shows_df['rounded_imdb_score'] = new_shows_df['imdb_score'].round()# rounding column with scores

print(new_shows_df.tail())#checking the outcome with tail()


                     name     character      role    title       type  \
85573       A??da Morales       Maritza     ACTOR  Lokillo  the movie   
85574    Adelaida Buscato    Mar??a Paz     ACTOR  Lokillo  the movie   
85575  Luz Stella Luengas  Karen Bayona     ACTOR  Lokillo  the movie   
85576        In??s Prieto         Fanny     ACTOR  Lokillo  the movie   
85578      Julian Gaviria       unknown  DIRECTOR  Lokillo  the movie   

       release_year      genres  imdb_score  imdb_votes  rounded_imdb_score  
85573          2021  ['comedy']         3.8        68.0                 4.0  
85574          2021  ['comedy']         3.8        68.0                 4.0  
85575          2021  ['comedy']         3.8        68.0                 4.0  
85576          2021  ['comedy']         3.8        68.0                 4.0  
85578          2021  ['comedy']         3.8        68.0                 4.0  


/tmp/ipykernel_47/3653630023.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_shows_df['rounded_imdb_score'] = new_shows_df['imdb_score'].round()# rounding column with scores


In [22]:
votes_count_by_score = new_shows_df.groupby('rounded_imdb_score')['imdb_votes'].nunique()
print(votes_count_by_score)# Use groupby() for scores and count all unique values in each group, print the result


rounded_imdb_score
2.0       20
3.0       51
4.0      208
5.0      530
6.0     1331
7.0     1389
8.0      865
9.0       79
10.0       1
Name: imdb_votes, dtype: int64


Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [23]:
filtered_df = new_shows_df[(new_shows_df['rounded_imdb_score'] >= 4) & (new_shows_df['rounded_imdb_score'] <= 9)]# filter dataframe using two conditions (scores to be in the range 4-9)

average_votes_by_score = filtered_df.groupby('rounded_imdb_score')['imdb_votes'].mean().reset_index()# group scores and corresponding average number of votes, reset index and print the result

print(average_votes_by_score)

   rounded_imdb_score     imdb_votes
0                 4.0    8699.529820
1                 5.0   14236.204123
2                 6.0   28807.336192
3                 7.0   51308.138542
4                 8.0  145532.023081
5                 9.0  251330.439834


In [24]:
average_votes_by_score['rounded_average_imdb_votes'] = average_votes_by_score['imdb_votes'].round()# round column with averages

average_votes_by_score = average_votes_by_score.rename(columns={'rounded_imdb_score': 'Rounded Imdb Score', 'rounded_average_imdb_votes': 'Rounded Average Imdb Votes'})# rename columns

final_result = average_votes_by_score.sort_values(by='Rounded Average Imdb Votes', ascending=False)
print(final_result)# print dataframe in descending order


   Rounded Imdb Score     imdb_votes  Rounded Average Imdb Votes
5                 9.0  251330.439834                    251330.0
4                 8.0  145532.023081                    145532.0
3                 7.0   51308.138542                     51308.0
2                 6.0   28807.336192                     28807.0
1                 5.0   14236.204123                     14236.0
0                 4.0    8699.529820                      8700.0


The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.